#### Importamos Librerias genéricas:
    NumPy, Pandas, IPython.
#### Importamos el Dataset de Properati 
    Se importa como Dtype: string

In [1]:
import pandas as pd
import numpy as np
from IPython.display import display
data = pd.read_csv('properati.csv', dtype={'all': str})

Se toma una muestra del Dataframe original, para su visualización general.

In [2]:
data.head()

Unnamed: 0 operation property_type place_name  \
0           0      sell            PH  Mataderos   
1           1      sell     apartment   La Plata   
2           2      sell     apartment  Mataderos   
3           3      sell            PH    Liniers   
4           4      sell     apartment     Centro   

                             place_with_parent_names country_name  \
0              |Argentina|Capital Federal|Mataderos|    Argentina   
1        |Argentina|Bs.As. G.B.A. Zona Sur|La Plata|    Argentina   
2              |Argentina|Capital Federal|Mataderos|    Argentina   
3                |Argentina|Capital Federal|Liniers|    Argentina   
4  |Argentina|Buenos Aires Costa Atlántica|Mar de...    Argentina   

                     state_name  geonames_id                  lat-lon  \
0               Capital Federal    3430787.0  -34.6618237,-58.5088387   
1        Bs.As. G.B.A. Zona Sur    3432039.0  -34.9038831,-57.9643295   
2               Capital Federal    3430787.0  -34.6522615,-58.5229825   
3               Capital Federal    3431333.0  -34.6477969,-58.5164244   
4  Buenos Aires Costa Atlántica    3435548.0  -38.0026256,-57.5494468   

         lat  ...  surface_covered_in_m2  price_usd_per_m2 price_per_m2  \
0 -34.661824  ...                   40.0       1127.272727  1550.000000   
1 -34.903883  ...                    NaN               NaN          NaN   
2 -34.652262  ...                   55.0       1309.090909  1309.090909   
3 -34.647797  ...                    NaN               NaN          NaN   
4 -38.002626  ...                   35.0       1828.571429  1828.571429   

   floor  rooms  expenses                                      properati_url  \
0    NaN    NaN       NaN  http://www.properati.com.ar/15bo8_venta_ph_mat...   
1    NaN    NaN       NaN  http://www.properati.com.ar/15bob_venta_depart...   
2    NaN    NaN       NaN  http://www.properati.com.ar/15bod_venta_depart...   
3    NaN    NaN       NaN  http://www.properati.com.ar/15boh_venta_ph_lin...   
4    NaN    NaN       NaN  http://www.properati.com.ar/15bok_venta_depart...   

                                         description  \
0  2 AMBIENTES TIPO CASA PLANTA BAJA POR PASILLO,...   
1  Venta de departamento en décimo piso al frente...   
2  2 AMBIENTES 3ER PISO LATERAL LIVING COMEDOR AM...   
3  PH 3 ambientes con patio. Hay 3 deptos en lote...   
4  DEPARTAMENTO CON FANTÁSTICA ILUMINACIÓN NATURA...   

                                               title  \
0                 2 AMB TIPO CASA SIN EXPENSAS EN PB   
1  VENTA Depto 2 dorm. a estrenar 7 e/ 36 y 37   ...   
2           2 AMB 3ER PISO CON ASCENSOR APTO CREDITO   
3                       PH 3 amb. cfte. reciclado      
4  DEPTO 2 AMB AL CONTRAFRENTE ZONA CENTRO/PLAZA ...   

                                     image_thumbnail  
0  https://thumbs4.properati.com/8/BluUYiHJLhgIIK...  
1  https://thumbs4.properati.com/7/ikpVBu2ztHA7jv...  
2  https://thumbs4.properati.com/5/SXKr34F_IwG3W_...  
3  https://thumbs4.properati.com/3/DgIfX-85Mog5SP...  
4  https://thumbs4.properati.com/5/xrRqlNcSI_vs-f...  

[5 rows x 26 columns]

In [3]:
#Obtenemos información básica del dataframe, cantidad de datos vs. no nulos.
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121220 entries, 0 to 121219
Data columns (total 26 columns):
Unnamed: 0                    121220 non-null int64
operation                     121220 non-null object
property_type                 121220 non-null object
place_name                    121197 non-null object
place_with_parent_names       121220 non-null object
country_name                  121220 non-null object
state_name                    121220 non-null object
geonames_id                   102503 non-null float64
lat-lon                       69670 non-null object
lat                           69670 non-null float64
lon                           69670 non-null float64
price                         100810 non-null float64
currency                      100809 non-null object
price_aprox_local_currency    100810 non-null float64
price_aprox_usd               100810 non-null float64
surface_total_in_m2           81892 non-null float64
surface_covered_in_m2         101313 no

#### Armamos 2 DataFrame:
    (A) 1 DataFrame para Ordenada Latitud: de columnas 'place_name' y 'lat'
    (B) 1 DataFrame para Ordenada Longitud: de columnas 'place_name' y 'lon'

(A) DF de Ordenada Latitud

In [4]:
# Definimos el subset del dataframe 'data' para quedarnos sólo con las columnas 'place_name' y 'lat'
# una vez definida se pasa el método ".head()" que despliega las primeras 5 filas del Dataframe 

dflatitud=data[['place_name', 'lat']]
dflatitud.head()

place_name        lat
0  Mataderos -34.661824
1   La Plata -34.903883
2  Mataderos -34.652262
3    Liniers -34.647797
4     Centro -38.002626

Comprobamos cantidad de filas completas de la columna 'lat' en comparación al total de filas

In [5]:
dflatitud.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121220 entries, 0 to 121219
Data columns (total 2 columns):
place_name    121197 non-null object
lat           69670 non-null float64
dtypes: float64(1), object(1)
memory usage: 1.8+ MB


De (A) nos quedamos solo con los campos con datos, de la columna 'lat', desechando los valores 'NaN'

In [6]:
dflatitud_1=dflatitud.dropna()
dflatitud_1.head()

place_name        lat
0  Mataderos -34.661824
1   La Plata -34.903883
2  Mataderos -34.652262
3    Liniers -34.647797
4     Centro -38.002626

Volvemos a tomar cantidad de datos por fila y vemos que se eliminaron los NaN del Dataframe.
Observamos también, que en 'lat' bajaron 23 filas más, lo que indica que en 'place_name' también eran NaN. Se eliminan son sólo 23 en 121220 (0,018%)

In [7]:
dflatitud_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 69647 entries, 0 to 121217
Data columns (total 2 columns):
place_name    69647 non-null object
lat           69647 non-null float64
dtypes: float64(1), object(1)
memory usage: 1.6+ MB


Calulamos la media de las latitudes por grupo de localidad, con el método ".groupby()"
P.ej.: toma todas las filas de localidad igual a "Abasto", de la columna 'place_name' y calcula la media de las Latitudes

In [8]:
dflatitud_1=dflatitud_1.groupby(['place_name']).mean()
dflatitud_1.head()

lat
place_name                        
 Country Maschwitz Club -34.377707
 los alamos             -34.423989
9 de Abril              -34.756258
Abasto                  -34.639208
Abril Club de Campo     -34.814314

Comprobamos que la cantidad de filas vuelve a disminuir, puesto que sólo muestra la media de localidades únicas.

In [9]:
len(dflatitud_1)

959

______________________________________________________________________________________________________________________

(B) DF de Ordenada Longitud

#### Armo un subset sólo con las columnas 'place_name' y 'lon'

In [10]:
# Definimos el subset del dataframe 'data' para quedarnos sólo con las columnas 'place_name' y 'lat'
# una vez definida se pasa el método ".head()" que despliega las primeras 5 filas del Dataframe 

dflongitud=data[['place_name', 'lon']]
dflongitud.head()

place_name        lon
0  Mataderos -58.508839
1   La Plata -57.964330
2  Mataderos -58.522982
3    Liniers -58.516424
4     Centro -57.549447

Comprobamos cantidad de filas completas de la columna 'lon' en comparación al total de filas

In [11]:
dflongitud.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121220 entries, 0 to 121219
Data columns (total 2 columns):
place_name    121197 non-null object
lon           69670 non-null float64
dtypes: float64(1), object(1)
memory usage: 1.8+ MB


De (B) nos quedamos solo con los campos con datos, de la columna 'lon', desechando los valores 'NaN'

In [12]:
dflongitud_1=dflongitud.dropna()
dflongitud_1.head()

place_name        lon
0  Mataderos -58.508839
1   La Plata -57.964330
2  Mataderos -58.522982
3    Liniers -58.516424
4     Centro -57.549447

Volvemos a tomar cantidad de datos por fila y vemos que se eliminaron los NaN del Dataframe.
Observamos también, que en 'lon' bajaron 23 filas más, lo que indica que en 'place_name' también eran NaN. Se eliminan son sólo 23 en 121220 (0,018%)

In [13]:
dflongitud_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 69647 entries, 0 to 121217
Data columns (total 2 columns):
place_name    69647 non-null object
lon           69647 non-null float64
dtypes: float64(1), object(1)
memory usage: 1.6+ MB


Calulamos la media de las longtudes por grupo de localidad, con el método ".groupby()" P.ej.: toma todas las filas de localidad igual a "Abasto", de la columna 'place_name' y calcula la media de las Longitudes

In [14]:
dflongitud_1=dflongitud_1.groupby(['place_name']).mean()
dflongitud_1.head()

lon
place_name                        
 Country Maschwitz Club -58.755411
 los alamos             -58.601876
9 de Abril              -58.498288
Abasto                  -58.380827
Abril Club de Campo     -58.156066

Comprobamos que la cantidad de filas vuelve a disminuir, puesto que sólo muestra la media de localidades únicas.

In [15]:
len(dflongitud_1)

959

Definimos un nuevo dataframe que una las columnas lat y lon de los puntos (A) y (B).
Para ello vamos utilizar el método ".merge()" de Pandas

In [16]:
dflatlon=pd.merge(dflatitud_1, dflongitud_1, left_on='place_name', right_on='place_name')
dflatlon

lat        lon
place_name                                   
 Country Maschwitz Club -34.377707 -58.755411
 los alamos             -34.423989 -58.601876
9 de Abril              -34.756258 -58.498288
Abasto                  -34.639208 -58.380827
Abril Club de Campo     -34.814314 -58.156066
...                            ...        ...
William Morris          -34.581910 -58.655338
Yerba Buena             -26.813039 -65.294461
Zelaya                  -35.053791 -59.660925
Zárate                  -34.096436 -59.067763
coordenadas 34.255511   -34.428485 -58.601717

[959 rows x 2 columns]

Tomamos al nuevo dataframe y generamos un nuevo índice para que 'place_name' vuelva a ser una variable del df

In [17]:
dflatlon.reset_index(inplace=True)
dflatlon

place_name        lat        lon
0     Country Maschwitz Club -34.377707 -58.755411
1                 los alamos -34.423989 -58.601876
2                 9 de Abril -34.756258 -58.498288
3                     Abasto -34.639208 -58.380827
4        Abril Club de Campo -34.814314 -58.156066
..                       ...        ...        ...
954           William Morris -34.581910 -58.655338
955              Yerba Buena -26.813039 -65.294461
956                   Zelaya -35.053791 -59.660925
957                   Zárate -34.096436 -59.067763
958    coordenadas 34.255511 -34.428485 -58.601717

[959 rows x 3 columns]

Tomamos al dataframe original (data) y hacemos un Join con Pandas, sumando al final del DF las columnas con las medias de 'lat' y 'lon'

In [18]:
data_1= data.merge(dflatlon, how='left', left_on='place_name', right_on='place_name')
data_1

Unnamed: 0 operation property_type       place_name  \
0                0      sell            PH        Mataderos   
1                1      sell     apartment         La Plata   
2                2      sell     apartment        Mataderos   
3                3      sell            PH          Liniers   
4                4      sell     apartment           Centro   
...            ...       ...           ...              ...   
121215      121215      sell     apartment         Belgrano   
121216      121216      sell         house           Beccar   
121217      121217      sell     apartment    Villa Urquiza   
121218      121218      sell     apartment      Plaza Colón   
121219      121219      sell     apartment  Capital Federal   

                                  place_with_parent_names country_name  \
0                   |Argentina|Capital Federal|Mataderos|    Argentina   
1             |Argentina|Bs.As. G.B.A. Zona Sur|La Plata|    Argentina   
2                   |Argentina|Capital Federal|Mataderos|    Argentina   
3                     |Argentina|Capital Federal|Liniers|    Argentina   
4       |Argentina|Buenos Aires Costa Atlántica|Mar de...    Argentina   
...                                                   ...          ...   
121215               |Argentina|Capital Federal|Belgrano|    Argentina   
121216  |Argentina|Bs.As. G.B.A. Zona Norte|San Isidro...    Argentina   
121217          |Argentina|Capital Federal|Villa Urquiza|    Argentina   
121218  |Argentina|Buenos Aires Costa Atlántica|Mar de...    Argentina   
121219                        |Argentina|Capital Federal|    Argentina   

                          state_name  geonames_id  \
0                    Capital Federal    3430787.0   
1             Bs.As. G.B.A. Zona Sur    3432039.0   
2                    Capital Federal    3430787.0   
3                    Capital Federal    3431333.0   
4       Buenos Aires Costa Atlántica    3435548.0   
...                              ...          ...   
121215               Capital Federal    3436077.0   
121216      Bs.As. G.B.A. Zona Norte    3436080.0   
121217               Capital Federal    3433775.0   
121218  Buenos Aires Costa Atlántica          NaN   
121219               Capital Federal    3433955.0   

                              lat-lon      lat_x  ...  price_per_m2  floor  \
0             -34.6618237,-58.5088387 -34.661824  ...   1550.000000    NaN   
1             -34.9038831,-57.9643295 -34.903883  ...           NaN    NaN   
2             -34.6522615,-58.5229825 -34.652262  ...   1309.090909    NaN   
3             -34.6477969,-58.5164244 -34.647797  ...           NaN    NaN   
4             -38.0026256,-57.5494468 -38.002626  ...   1828.571429    NaN   
...                               ...        ...  ...           ...    ...   
121215                            NaN        NaN  ...   9354.838710    NaN   
121216                            NaN        NaN  ...   1383.333333    NaN   
121217  -34.5706388726,-58.4755963355 -34.570639  ...   3371.794872    NaN   
121218                            NaN        NaN  ...   1997.916667    NaN   
121219                            NaN        NaN  ...   1675.324675    NaN   

       rooms  expenses                                      properati_url  \
0        NaN       NaN  http://www.properati.com.ar/15bo8_venta_ph_mat...   
1        NaN       NaN  http://www.properati.com.ar/15bob_venta_depart...   
2        NaN       NaN  http://www.properati.com.ar/15bod_venta_depart...   
3        NaN       NaN  http://www.properati.com.ar/15boh_venta_ph_lin...   
4        NaN       NaN  http://www.properati.com.ar/15bok_venta_depart...   
...      ...       ...                                                ...   
121215   NaN   10000.0  http://www.properati.com.ar/1cja2_venta_depart...   
121216   NaN       NaN  http://www.properati.com.ar/1cja6_venta_casa_b...   
121217   NaN       NaN  http://www.properati.com.ar/1cja7_venta_depart...   
121218   NaN      

Volvermos a conformar un nuevo dataframe con las latitudes originales y las medias por grupo

In [19]:
data_1_Lats=data_1[['place_name','lat_y','lat_x']]
data_1_Lats

place_name      lat_y      lat_x
0             Mataderos -34.666184 -34.661824
1              La Plata -34.924338 -34.903883
2             Mataderos -34.666184 -34.652262
3               Liniers -34.643470 -34.647797
4                Centro -38.000247 -38.002626
...                 ...        ...        ...
121215         Belgrano -34.563541        NaN
121216           Beccar -34.462778        NaN
121217    Villa Urquiza -34.570370 -34.570639
121218      Plaza Colón -38.007186        NaN
121219  Capital Federal -34.626709        NaN

[121220 rows x 3 columns]

Hacemos lo mismo que en el paso anterior, para las columnas de longitud

In [20]:
data_1_Lons=data_1[['place_name','lon_y','lon_x']]
data_1_Lons

place_name      lon_y      lon_x
0             Mataderos -58.518986 -58.508839
1              La Plata -57.966073 -57.964330
2             Mataderos -58.518986 -58.522982
3               Liniers -58.516909 -58.516424
4                Centro -57.547260 -57.549447
...                 ...        ...        ...
121215         Belgrano -58.453904        NaN
121216           Beccar -58.533655        NaN
121217    Villa Urquiza -58.484773 -58.475596
121218      Plaza Colón -57.543931        NaN
121219  Capital Federal -58.435882        NaN

[121220 rows x 3 columns]

Extendemos con método '.fillna()' las medias por localidad la columna con latitudes NaN, con el fin de completar todas las filas con datos

In [25]:
data_1_Lats_def=data_1_Lats.fillna(axis=1, method='pad')
data_1_Lats_def

place_name    lat_y    lat_x
0             Mataderos -34.6662 -34.6618
1              La Plata -34.9243 -34.9039
2             Mataderos -34.6662 -34.6523
3               Liniers -34.6435 -34.6478
4                Centro -38.0002 -38.0026
...                 ...      ...      ...
121215         Belgrano -34.5635 -34.5635
121216           Beccar -34.4628 -34.4628
121217    Villa Urquiza -34.5704 -34.5706
121218      Plaza Colón -38.0072 -38.0072
121219  Capital Federal -34.6267 -34.6267

[121220 rows x 3 columns]

In [29]:
dfLatitud_definitva=data_1_Lats_def[['place_name','lat_x']]
dfLatitud_definitva

place_name    lat_x
0             Mataderos -34.6618
1              La Plata -34.9039
2             Mataderos -34.6523
3               Liniers -34.6478
4                Centro -38.0026
...                 ...      ...
121215         Belgrano -34.5635
121216           Beccar -34.4628
121217    Villa Urquiza -34.5706
121218      Plaza Colón -38.0072
121219  Capital Federal -34.6267

[121220 rows x 2 columns]

Extendemos con método '.fillna()' las medias por localidad la columna con longitudes NaN, con el fin de completar todas las filas con datos

In [96]:
data_1_Lons_def=data_1_Lons.fillna(axis=1, method='pad')
data_1_Lons_def

place_name    lon_y    lon_x
0             Mataderos  -58.519 -58.5088
1              La Plata -57.9661 -57.9643
2             Mataderos  -58.519  -58.523
3               Liniers -58.5169 -58.5164
4                Centro -57.5473 -57.5494
...                 ...      ...      ...
121215         Belgrano -58.4539 -58.4539
121216           Beccar -58.5337 -58.5337
121217    Villa Urquiza -58.4848 -58.4756
121218      Plaza Colón -57.5439 -57.5439
121219  Capital Federal -58.4359 -58.4359

[121220 rows x 3 columns]

In [30]:
dfLongitud_definitva=data_1_Lons_def[['place_name','lon_x']]
dfLongitud_definitva

place_name    lon_x
0             Mataderos -58.5088
1              La Plata -57.9643
2             Mataderos  -58.523
3               Liniers -58.5164
4                Centro -57.5494
...                 ...      ...
121215         Belgrano -58.4539
121216           Beccar -58.5337
121217    Villa Urquiza -58.4756
121218      Plaza Colón -57.5439
121219  Capital Federal -58.4359

[121220 rows x 2 columns]